packages/core/src/actions/emit.ts

```diff
@@ -0,0 +1,100 @@
+import isDevelopment from '#is-development';
+import {
+  ActionArgs,
+  AnyActorScope,
+  AnyMachineSnapshot,
+  EventObject,
+  MachineContext,
+  SendExpr,
+  ParameterizedObject
+} from '../types.ts';
+
+function resolveEmit(
+  _: AnyActorScope,
+  snapshot: AnyMachineSnapshot,
+  args: ActionArgs<any, any, any>,
+  actionParams: ParameterizedObject['params'] | undefined,
+  {
+    event: eventOrExpr
+  }: {
+    event:
+      | EventObject
+      | SendExpr<
+          MachineContext,
+          EventObject,
+          ParameterizedObject['params'] | undefined,
+          EventObject,
+          EventObject
+        >;
+  }
+) {
+  if (isDevelopment && typeof eventOrExpr === 'string') {
+    throw new Error(
+      `Only event objects may be used with emit; use emit({ type: "${eventOrExpr}" }) instead`
+    );
+  }
+  const resolvedEvent =
+    typeof eventOrExpr === 'function'
+      ? eventOrExpr(args, actionParams)
+      : eventOrExpr;
+  return [snapshot, { event: resolvedEvent }];
+}
+
+function executeEmit(
+  actorScope: AnyActorScope,
+  params: {
+    event: EventObject;
+  }
+) {
+  const { event } = params;
+  actorScope.defer(() => {
```

Comment:
can we think of any sensible test case that would require `defer` here? I think that it's probably the right thing to use `defer` here - it just would be cool to have a test case that would prove its necessity

Reply:
There is no need; just removed it. It was a copypasta

Reply:
What snapshot should be read here? should it be the one from *before* or *after* the transition? I think making it the one *after* would be more consistent with `send`. We need `defer` for this and a test could be somewhat easily crafted for this situation
```ts
actorRef.on('foo', () => {
  actorRef.getSnapshot()
})
```

Reply:
Can you write a test for that?

Reply:
Added a test for this; it works without `defer`

Reply:
JK it doesn't